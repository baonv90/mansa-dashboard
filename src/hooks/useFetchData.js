import { useState, useEffect } from 'react';

// A hook to fetch data from given url
// it returns the fetching status loading and data
export const useFetchData = (url) => {
  const [data, setData] = useState();
  const [loading, setLoading] = useState(false);
  const [hasError, setHasError] = useState(false);

  useEffect(() => {
    setLoading(true);
    fetch(url).then((res) => {
      res.json().then((data) => {
        setData(data);
        setLoading(false);
      })
      .catch(() => {
        setHasError(true);
        setLoading(false);
      });
    });
  }, [url]);

  return { data, loading, hasError };
};