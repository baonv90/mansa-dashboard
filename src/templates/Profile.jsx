import React from 'react';
import styled from 'styled-components';
import get from 'lodash/get';


import { useFetchData } from '../hooks/useFetchData';
const data_source = 'https://randomuser.me/api/';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const Picture = styled.img`
  width: 100px;
  height: 100px;
  border-radius: 50%;
  margin-bottom: 1rem;
  border: 3px solid var(--border-color);
`;

export const TitleName = styled.span`
  font-size: 1.1rem;
  font-weight: 600;
  line-height: 1.5rem;
  margin-bottom: .5rem;
`;

export const Info = styled.span`
  font-size: .9rem;
  font-weight: 400;
  line-height: 1.25rem;
`;

// User profile from randomuser
const Profile = () => {
  const { data } = useFetchData(data_source);
  const userInfo = get(data, 'results[0]');

  const renderProfile = ({ phone, email, name, picture } = {}) => {
    const { first, last } = name;
    const displayName = first + ' ' + last;
    return <>
      <Picture src={picture ? picture.medium : ''} />
      <TitleName>{displayName}</TitleName>
      <Info>{email}</Info>
      <Info>{phone}</Info>
    </>
  };

  if (!userInfo) {
    return null;
  }

  return (
    <Container>
      {renderProfile(userInfo)}
    </Container>
  )
};

export default Profile;