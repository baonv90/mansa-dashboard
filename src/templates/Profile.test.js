import renderer from 'react-test-renderer';
import Profile from './Profile';
import { TitleName } from './Profile';
import { useFetchData } from '../hooks/useFetchData';

jest.mock('../hooks/useFetchData');

describe('templates > Company', () => {
  const fakeData = {
    results: [{
      name: {
        first: 'Goku',
        last: 'Son'
      },
      phone: '123456789',
      email: 'Gohan@Chichi.kame'
    }]
  };

  beforeEach(() => {
    useFetchData.mockReturnValue({
      loading: false,
      data: fakeData
    });
  });

  it('should match snapshot', async () =>  {
    const component = await renderer.create(<Profile />);
    
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should pass data correctly', async () =>  {
    const tree = await renderer.create(<Profile />);
    
    const title = tree.root.findByType(TitleName);
    expect(title.props.children).toEqual('Goku Son');    
  });
});
