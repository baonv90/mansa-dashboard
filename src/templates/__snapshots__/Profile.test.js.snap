// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates > Company should match snapshot 1`] = `
<div
  className="sc-bdnxRM cZDZdT"
>
  <img
    className="sc-gtsrHT cHsyZy"
    src=""
  />
  <span
    className="sc-dlnjwi emJSdu"
  >
    Goku Son
  </span>
  <span
    className="sc-hKFxyN hpHuXd"
  >
    Gohan@Chichi.kame
  </span>
  <span
    className="sc-hKFxyN hpHuXd"
  >
    123456789
  </span>
</div>
`;
